cli/azd/ci-build.ps1

```diff
@@ -39,43 +39,90 @@ if ($IsWindows) {
     Write-Host "go generate succeeded"
 }
 
+# Run `go help build` to obtain detailed information about `go build` flags.
+$buildFlags = @(
+    # Remove file system paths from the compiled binary
+    "-gcflags=-trimpath",
+    # Remove file system paths from the assembled code
+    "-asmflags=-trimpath",
+    # remove all file system paths from the resulting executable.
+    # Instead of absolute file system paths, the recorded file names
+    # will begin either a module path@version (when using modules),
+    # or a plain import path (when using the standard library, or GOPATH).
+    "-trimpath"
+)
+
+# Build constraint tags
+# cfi: Enable Control Flow Integrity (CFI),
+# cfg: Enable Control Flow Guard (CFG),
+# osusergo: Optimize for OS user accounts
+$tagsFlag = "-tags=cfi,cfg,osusergo"
+
+# ld linker flags
+# -s: Omit symbol table and debug information
+# -w: Omit DWARF symbol table
+# -X: Set variable at link time. Used to set the version in source.
+$ldFlag = "-ldflags=`"-s -w -X 'github.com/azure/azure-dev/cli/azd/internal.Version=$Version (commit $SourceVersion)' "
+
 if ($IsWindows) {
-    Write-Host "go build (windows)"
-    go build `
-        -buildmode=exe `
-        -tags="cfi,cfg,osusergo" `
-        -trimpath `
-        -gcflags="-trimpath" `
-        -asmflags="-trimpath" `
-        -ldflags="-s -w -X 'github.com/azure/azure-dev/cli/azd/internal.Version=$Version (commit $SourceVersion)' -linkmode=auto -extldflags=-Wl,--high-entropy-va"
+    Write-Host "Building for windows"
+    $buildFlags += @(
+        "-buildmode=exe",
+        $tagsFlag,
+        # -extldflags=-Wl,--high-entropy-va: Pass the high-entropy VA flag to the linker to enable high entropy virtual addresses
+        ($ldFlag + "-linkmode=auto -extldflags=-Wl,--high-entropy-va`"")
+    )
 }
 elseif ($IsLinux) {
-    Write-Host "go build (linux)"
-    go build `
-        -buildmode=pie `
-        -tags="cfi,cfg,cfgo,osusergo" `
-        -gcflags="-trimpath" `
-        -asmflags="-trimpath" `
-        -ldflags="-s -w -X 'github.com/azure/azure-dev/cli/azd/internal.Version=$Version (commit $SourceVersion)' -extldflags=-Wl,--high-entropy-va"
+    Write-Host "Building for linux"
+    $buildFlags += @(
```

Comment:
@wbreza in original [comment](https://github.com/Azure/azure-dev/pull/2007#discussion_r1177177253):
I found it much easier to read & understand the OS specific build command/flags in use prior to this change. This update requires additional cognitive load to decipher what should ideally be a simple build command.

Reply:
I agree completely with your observation and share the same concerns. I don't like it either. However, it avoids code duplication and may work out better in practice, to avoid cases like `trimpath` being shared incorrectly. That was the motivation here. Perhaps there's something better, and I'd be happy to take any suggestions.

To make things easier, the script will print out what `go build` command it generated. This can be observed both locally and in CI to verify that things are working when making changes, to verify that things work as expected. You can see the result of this [here](https://github.com/Azure/azure-dev/pull/2005/files#r1174041133).